File: src/pooh3/mobi/app/a01_clock/model/Degree.java
package pooh3.mobi.app.a01_clock.model;

import static pooh3.mobi.Preconditions.checkNotNull;

public class Degree {

    public final int val;

    private Degree(int val) {
        this.val = val;
    }

    public static Degree of(int degreeInt) {
        return new Degree(degreeInt);
    }

    public Degree minus(Degree degree) {
        return Degree.of(this.val - checkNotNull(degree).val);
    }

    public Degree minusAbs(Degree degree) {
        return Degree.of(
                Math.abs(
                        minus(checkNotNull(degree)).val));
    }

    public Degree minusMin(Degree degree) {
        final Degree calced = checkNotNull(minusAbs(degree));
        return calced.val <= 180 ? calced : Degree.of(360).minus(calced);
    }
}

File: src/pooh3/mobi/app/a01_clock/model/Degree.kt
package pooh3.mobi.app.a01_clock.model

import pooh3.mobi.checkNotNull


class Degree(val int: Int) {

    operator fun plus(degree: Degree): Degree {
        return Degree(this.int + degree.checkNotNull.int)
    }

    operator fun minus(degree: Degree): Degree {
        return Degree(this.int - degree.checkNotNull.int)
    }

    override fun toString(): String {
        return "[$int degree]"
    }

}

fun Degree.minusAbs(degree: Degree): Degree {
    return (this - degree.checkNotNull).absDegree
}

fun Degree.minusMin(degree: Degree): Degree {
    return this.minusAbs(degree.checkNotNull).let {
        return@let if (it.int <= 180) it else 360.degree.minus(it)
    }
}

val Int.degree: Degree
    get() = Degree(this)

val Degree.absDegree: Degree
    get() = Math.abs(this.int).degree